include/execution.hpp

```diff
@@ -2852,7 +2852,217 @@ namespace _P2300::execution {
+          __value_types_of_t<
+            _Sender,
+            make_env_t<with_t<get_stop_token_t, in_place_stop_token>>,
+            __mbind_front_q<__decayed_tuple, set_value_t>,
+            __q<__bind_tuples>>;
+
+        using __variant_t =
+          __error_types_of_t<
+            _Sender,
+            make_env_t<with_t<get_stop_token_t, in_place_stop_token>>,
+            __transform<
+              __mbind_front_q<__decayed_tuple, set_error_t>,
+              __bound_values_t>>;
+
+        using __receiver = __receiver<__sh_state>;
+
+        __variant_t __data_;
+        in_place_stop_source __stop_source_{};
+
+        std::atomic<void*> __op_state1_;
+        connect_result_t<_Sender, __receiver> __op_state2_;
+
+        explicit __sh_state(_Sender& __sndr)
+          : __op_state1_{nullptr}
+          , __op_state2_(connect((_Sender&&) __sndr, __receiver{*this}))
+        {
+          start(__op_state2_);
+        }
+
+        void __notify() noexcept {
+          void* __old{nullptr};
+          void* const __completion_state = static_cast<void*>(this);
+          bool const changed = __op_state1_.compare_exchange_weak(
+            __old, __completion_state,
+            std::memory_order_release,
+            std::memory_order_acquire
+          );
+          if(changed && __old != nullptr) {
+            static_cast<__operation_base*>(__old)->__notify_(static_cast<__operation_base*>(__old));
+          }
+        }
+      };
+
+    template <class _SenderId, class _ReceiverId>
+      class __operation : public __operation_base {
+        using _Sender = __t<_SenderId>;
+        using _Receiver = __t<_ReceiverId>;
+
+        struct __on_stop_requested {
+          in_place_stop_source& __stop_source_;
+          void operator()() noexcept {
+            __stop_source_.request_stop();
+          }
+        };
+        using __on_stop = std::optional<typename stop_token_of_t<
+            env_of_t<_Receiver> &>::template callback_type<__on_stop_requested>>;
+
+        _Receiver __recvr_;
+        __on_stop __on_stop_{};
+        std::shared_ptr<__sh_state<_SenderId>> __shared_state_;
+
+      public:
+        __operation(_Receiver&& __rcvr,
+                    std::shared_ptr<__sh_state<_SenderId>> __shared_state)
+            noexcept(std::is_nothrow_move_constructible_v<_Receiver>)
+          : __operation_base{__notify}
+          , __recvr_((_Receiver&&)__rcvr)
+          , __shared_state_(move(__shared_state)) {
+        }
+        _P2300_IMMOVABLE(__operation);
+
+        static void __notify(__operation_base* __self) noexcept {
+          __operation *__op = static_cast<__operation*>(__self);
+          __op->__on_stop_.reset();
+
+          std::visit([&](const auto& __tupl) noexcept -> void {
+            std::apply([&](auto __tag, const auto&... __args) noexcept -> void {
+              __tag((_Receiver&&) __op->__recvr_, __args...);
+            }, __tupl);
+          }, __op->__shared_state_->__data_);
```

Comment:
The intention is for `ensure_started` to return a single-shot sender (a sender that can only ever be connected to a receiver once). Because of that, the `ensure_started` spec requires that these values be moved in:

```suggestion
          std::visit([&](auto& __tupl) noexcept -> void {
            std::apply([&](auto __tag, auto&... __args) noexcept -> void {
              __tag((_Receiver&&) __op->__recvr_, std::move(__args)...);
            }, __tupl);
          }, __op->__shared_state_->__data_);
```